mainloop.cpp

```diff
@@ -59,6 +59,8 @@ int Mainloop::open()
         return -1;
     }
 
+    Endpoint::set_Mainloop(this);
```

Comment:
And yes... this means it's probably a good idea to move the call to Mainloop::open() to happen earlier, so we only create endpoints when/if we have a valid Mainloop instance